### PR TITLE
Add External IDs to Issues and Code to Projects

### PR DESCRIPTION
**Pull request by GhiocelAndrei**
+ Also added Column IssueSequence to Projects for Sequence Generation
Solves Issue #28

### CODE REVIEW

File: IssueTrackerApplication/Services/IssuesService.cs [hunk 1]
```diff
@@ -37,6 +39,17 @@ public override async Task<Issue> CreateAsync(CreateIssueCommand entity, Cancell
                 throw new InvalidInputException("Assignee ID does not exits");
             }
 
+            using var transaction = DbContext.Database.BeginTransaction();
+
+            await DbContext.Database.ExecuteSqlAsync($"UPDATE Projects WITH (HOLDLOCK, ROWLOCK) SET IssueSequence = IssueSequence + 1 WHERE Id = {entity.ProjectId}");
+            var returnedRow = DbContext.Projects.FromSql($"SELECT * FROM Projects WHERE Id = {entity.ProjectId}").FirstOrDefault();
```
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerApplication/Services/IssuesService.cs]: Both comments solved in the latest commit

File: IssueTracker.Abstractions/Definitions/IssueSequenceReturn.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+﻿namespace IssueTracker.Abstractions.Definitions
+{
+    public class IssueSequenceReturn
```
Comment (cristi-paraschiv) [inline IssueTracker.Abstractions/Definitions/IssueSequenceReturn.cs:3]: I would generalize this and name it something like `ScalarLong`; this way you could reuse it in the future

File: IssueTrackerDataAccess/Repositories/IProjectRepository.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿namespace IssueTracker.DataAccess.Repositories
+{
+    public interface IProjectRepository
+    {
+        long UpdateAndGetIssueSequence(long projectId);
```
Comment (cristi-paraschiv) [inline IssueTrackerDataAccess/Repositories/IProjectRepository.cs:5]: use async and let's name it just `GetIssueSequence`

File: IssueTrackerDataAccess/Repositories/ProjectRepository.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        public ProjectRepository(IssueContext dbContext)
+        {
+            _dbContext = dbContext;
+        }
+
+        public long UpdateAndGetIssueSequence(long projectId)
+        {
+            var projectIdParameter = new SqlParameter("@ProjectId", projectId);
+
+            var sqlQuery = "UPDATE Projects WITH (HOLDLOCK, ROWLOCK) SET IssueSequence = IssueSequence + 1 WHERE Id = @ProjectId;" +
```
Comment (cristi-paraschiv) [inline IssueTrackerDataAccess/Repositories/ProjectRepository.cs:20]: you can do this in a single statement

```
UPDATE Projects WITH (HOLDLOCK, ROWLOCK)
SET IssueSequence = IssueSequence + 1
OUTPUT inserted.IssueSequence
WHERE Id = @ProjectId
```

File: IssueTrackerDataAccess/Repositories/ProjectRepository.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    public class ProjectRepository : IProjectRepository
+    {
+        private readonly IssueContext _dbContext;
+
+        public ProjectRepository(IssueContext dbContext)
+        {
+            _dbContext = dbContext;
+        }
+
+        public long UpdateAndGetIssueSequence(long projectId)
```
Comment (cristi-paraschiv) [inline IssueTrackerDataAccess/Repositories/ProjectRepository.cs:16]: use async

File: IssueTrackerDataAccess/Migrations/AddColumnsForExternalIdAndProjectCode.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
+﻿using FluentMigrator;
+
+namespace IssueTracker.DataAccess.Migrations
+{
+    [Migration(202308011742)]
+    public class AddColumnsForExternalIdAndProjectCode : Migration
+    {
+        public override void Up()
+        {
+            Alter.Table("Issues")
```
Comment (cristi-paraschiv) [inline IssueTrackerDataAccess/Migrations/AddColumnsForExternalIdAndProjectCode.cs:10]: imagine this is an app that has been already running in the cloud for a few months; after we deploy this version some issues will have external ID, some won't - this is an inconsitent state, we need to fix this by generating them in a migration;

we could defer the population of this column for a future PR, where we actually use them (please create a ticket)

File: IssueTrackerDataAccess/IssueTracker.DataAccess.csproj [hunk 7]
```diff
@@ -13,6 +13,7 @@
     <PackageReference Include="Microsoft.AspNetCore.OpenApi" Version="7.0.9" />
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="7.0.9" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="7.0.9" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="7.0.9" />
```
Comment (cristi-paraschiv) [inline IssueTrackerDataAccess/IssueTracker.DataAccess.csproj:16]: is this really needed?

File: IssueTracker.ApplicationTesting/ServicesTest/IssueServiceTests.cs [hunk 8]
```diff
@@ -171,6 +171,7 @@ public async Task Create_ShouldThrow_WhenNonExistingUser()
                 await _sut.CreateAsync(createIssueCommand, It.IsAny<CancellationToken>()));
         }
 
+        /*
```
Comment (cristi-paraschiv) [inline IssueTracker.ApplicationTesting/ServicesTest/IssueServiceTests.cs:174]: why are these commented-out?

File: IssueTrackerDataAccess/Repositories/IProjectRepository.cs (hunk 3 above)
Comment (florincatrina) [reply IssueTrackerDataAccess/Repositories/IProjectRepository.cs:5]: and add cancellation token

File: IssueTracker.Abstractions/Definitions/IssueSequenceReturn.cs (hunk 2 above)
Comment (GhiocelAndrei) [reply IssueTracker.Abstractions/Definitions/IssueSequenceReturn.cs]: Renamed class in the latest commit

File: IssueTrackerDataAccess/Repositories/IProjectRepository.cs (hunk 3 above)
Comment (GhiocelAndrei) [reply IssueTrackerDataAccess/Repositories/IProjectRepository.cs]: I changed 'UpdateAndGetIssueSequence' to 'GetIssueSequenceAsync' function

File: IssueTrackerDataAccess/Repositories/ProjectRepository.cs (hunk 4 above)
Comment (GhiocelAndrei) [reply IssueTrackerDataAccess/Repositories/ProjectRepository.cs]: Rewrote query in the latest commit.

File: IssueTrackerDataAccess/Migrations/AddColumnsForExternalIdAndProjectCode.cs (hunk 6 above)
Comment (GhiocelAndrei) [reply IssueTrackerDataAccess/Migrations/AddColumnsForExternalIdAndProjectCode.cs:10]: I created Issue #37 in which I will solve this issue

File: IssueTrackerDataAccess/Repositories/ProjectRepository.cs (hunk 5 above)
Comment (GhiocelAndrei) [reply IssueTrackerDataAccess/Repositories/ProjectRepository.cs]: Solved in the latest commit.

File: IssueTracker.ApplicationTesting/ServicesTest/IssueServiceTests.cs (hunk 8 above)
Comment (GhiocelAndrei) [reply IssueTracker.ApplicationTesting/ServicesTest/IssueServiceTests.cs]: The test was throwing an Exception, so I assumed it was because we were using a TransactionScope, but after I had a second look I realized I wasn't mocking 'GetIssueSequenceAsync' and 'GetProjectCodeAsync' functions, but I fixed the test now

File: IssueTrackerDataAccess/Migrations/AddColumnsForExternalIdAndProjectCode.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+                 .AddColumn("ExternalId").AsString().Nullable();
+
+            Alter.Table("Projects")
+                 .AddColumn("Code").AsString().Nullable()
+                 .AddColumn("IssueSequence").AsInt64().NotNullable().WithDefaultValue(0L);
```
Comment (florincatrina) [inline IssueTrackerDataAccess/Migrations/AddColumnsForExternalIdAndProjectCode.cs:15]: As a design principle I would have used a dedicated table for sequences. Generating sequence creates a lot of contention on that Project record and might block queries that are not related. (just an fyi there is a sequence object that  you could have used but this approach is more didactical)

**Review by florincatrina — APPROVED** [review]

File: IssueTrackerApplication/Services/IssuesService.cs [hunk 10]
```diff
@@ -5,14 +5,16 @@
 using IssueTracker.Abstractions.Exceptions;
 using Microsoft.EntityFrameworkCore;
 using System.Transactions;
+using Microsoft.Data.SqlClient;
+using Microsoft.EntityFrameworkCore.Metadata.Internal;
```
Comment (cristi-paraschiv) [inline IssueTrackerApplication/Services/IssuesService.cs:9]: please clean unused usings

**Review by cristi-paraschiv — APPROVED** [review]
please squash commits before merging

File: IssueTrackerDataAccess/Migrations/AddColumnsForExternalIdAndProjectCode.cs (hunk 9 above)
Comment (GhiocelAndrei) [reply IssueTrackerDataAccess/Migrations/AddColumnsForExternalIdAndProjectCode.cs:15]: I will look into what sequence objects are

File: IssueTrackerApplication/Services/IssuesService.cs (hunk 10 above)
Comment (GhiocelAndrei) [reply IssueTrackerApplication/Services/IssuesService.cs]: Cleaned unused usings
